activation/activation_test.go

```diff
@@ -114,233 +132,210 @@ func (*ValidatorMock) Validate(nipst *types.NIPST, expectedChallenge common.Hash
-	chlng := common.HexToHash("0x3333")
-	poetRef := []byte{0xbe, 0xef}
-	nipstBuilder.SetPoetRef(poetRef)
-	npst := nipst.NewNIPSTWithChallenge(&chlng, poetRef)
-
-	atx := types.NewActivationTx(types.NodeId{"aaaaaa", []byte("bbbbb")}, coinbase, 1, prevAtx, 5, 1, prevAtx, 5, []types.BlockID{1, 2, 3}, npst)
+// ========== Tests ==========
 
-	err := activationDb.StoreAtx(atx.PubLayerIdx.GetEpoch(layersPerEpoch), atx)
-	assert.NoError(t, err)
+func TestBuilder_PublishActivationTx_HappyFlow(t *testing.T) {
+	r := require.New(t)
 
-	challenge := types.NIPSTChallenge{
-		NodeId:         b.nodeId,
-		Sequence:       b.GetLastSequence(b.nodeId) + 1,
-		PrevATXId:      atx.Id(),
-		PubLayerIdx:    atx.PubLayerIdx.Add(b.layersPerEpoch),
-		StartTick:      atx.EndTick,
-		EndTick:        b.tickProvider.NumOfTicks(), //todo: add provider when
-		PositioningAtx: atx.Id(),
-	}
+	// setup
+	activationDb := newActivationDb()
+	b := newBuilder(activationDb)
+	setActivesetSizeInCache(t, defaultActiveSetSize)
 
-	bytes, err := challenge.Hash()
-	npst2 := nipst.NewNIPSTWithChallenge(bytes, poetRef)
-	assert.NoError(t, err)
+	prevAtx := newAtx(newChallenge(nodeId, 1, prevAtxId, prevAtxId, 5), 5, []types.BlockID{1, 2, 3}, npst)
+	err := activationDb.StoreAtx(prevAtx.PubLayerIdx.GetEpoch(layersPerEpoch), prevAtx)
+	r.NoError(err)
 
-	//ugly hack to set view size in db
-	view, err := layers.GetOrphanBlocksBefore(layers.LatestLayer())
-	assert.NoError(t, err)
-	v, err := types.ViewAsBytes(view)
-	assert.NoError(t, err)
-	activesetCache.put(common.BytesToHash(v), 10)
+	// create and publish ATX
+	_, err = b.PublishActivationTx(meshProvider.LatestLayer().GetEpoch(layersPerEpoch))
+	r.NoError(err)
 
-	view, err = b.mesh.GetOrphanBlocksBefore(layers.LatestLayer())
-	assert.NoError(t, err)
-	activeSetSize, err := b.activeSet.ActiveSetSize(1)
-	assert.NoError(t, err)
-	act := types.NewActivationTx(b.nodeId, coinbase, b.GetLastSequence(b.nodeId)+1, atx.Id(), atx.PubLayerIdx+10, 0, atx.Id(), activeSetSize, view, npst2)
-	_, err = b.PublishActivationTx(1)
-	assert.NoError(t, err)
+	// assert that last transmission matches the ATX we expect
+	challenge := newChallenge(nodeId, prevAtx.Sequence+1, prevAtx.Id(), prevAtx.Id(), prevAtx.PubLayerIdx.Add(b.layersPerEpoch))
+	challengeHash, err := challenge.Hash()
+	r.NoError(err)
+	newNipst := nipst.NewNIPSTWithChallenge(challengeHash, poetRef)
 
-	bts, err := types.AtxAsBytes(act)
-	assert.NoError(t, err)
-	assert.Equal(t, bts, net.bt)
-	assert.Equal(t, bytes, nipstBuilder.Challenge)
+	expectedAtxBytes, err := types.AtxAsBytes(newAtx(challenge, defaultActiveSetSize, defaultView, newNipst))
+	r.NoError(err)
+	r.Equal(expectedAtxBytes, net.lastTransmission)
 
-	//test publish 2 transaction in same epoch
-	published, err := b.PublishActivationTx(2)
-	assert.False(t, published)
+	// assert that the challenge hash was set in the nipst builder
+	r.Equal(challengeHash, nipstBuilder.Challenge)
+
+	// cleanup
+	activesetCache.Purge()
```

Comment:
do you check publishing 2 transactions in the same epoch?

Reply:
[Yes](https://github.com/spacemeshos/go-spacemesh/pull/1188/files#diff-8523c0e2b5510da625a26461aa97dc46R259).